### PR TITLE
DOC: Improve `numpy.memmap` docs.

### LINKED ISSUE #12926: Add npy support to np.memmap
### Reproducing code example:

Here are to ways of creating a new empty array from a complex dtype object, if method 1 is used it creates a raw binary array on disk without the description header that np.save creates, were as method 2 first creates an empty array then saves it to disk and finally reloads it as a writable memmap'ed file, this is inefficient.

Under most circumstances this is fine as we would populate the array then save it to disk, however consider the example dtype this represents the results of a numerical simulation with two variables  hence each result is calculated in a loop over these variables, we could save the data in separate files but this messes up the file system.

The problem is when the simulation takes several hours to run if we wait to save we could lose all our results to a crash or force termination on a PBS system, were as with the memmap'ed method we can include a flag to indicate that a result has been calculated this way in the event of a crash of termination, we can resume without losing the results that had already been calculated. 

```python
import numpy as np
var1 = range(10)
var2 = range(20)
dtype= np.dtype([
('param1', float),
('param2', float),
('param3', float),
('param4', float, (2,)),
('var1', float, (len(var1),)),
('var2', float, (len(var2),)),
('result1', float, (len(var1),len(var2))),
('result2', float, (len(var1),len(var2),5,60))
])

#Method 1
data_memmap = np.memmap("memmap.npy", dtype, "w+", shape=())

#Method 2
data = np.empty((), dtype)
np.save("save.npy", data)
data_save = np.load("save.npy", "r+")
```

On my computer using method 1 has no header where as method 2 has the header 
```
{
  'descr': [('param1', '<f8'),
               ('param2', '<f8'),
               ('param3', '<f8'),
               ('param4', '<f8', (2,)),
               ('var1', '<f8', (10,)),
               ('var2', '<f8', (20,)),
               ('result1', '<f8', (10, 20)),
               ('result2', '<f8', (10, 20, 5, 60))],
 'fortran_order': False,
 'shape': (),
}
```

both allow me to access the array as a memmap'ed file which can be flushed using `data_x.flush()`

What I want is a combination method say

```python
data_memmap = np.memmap("memmap.npy", dtype, "w+", shape=(), npy=True)
```

or

```python
data_memmap = np.memmap("memmap.npy", dtype, "nw+", shape=())
```

which work like the normal memmap command but removes the offset option and creates a new npy file to map to, this avoids the inefficiency of method 2.

### Numpy/Python version information:

1.14.5 3.7.0 (v3.7.0:1bf9cc5093, Jun 27 2018, 04:59:51) [MSC v.1914 64 bit (AMD64)]

### PR DESCRIPTION
**Pull request by lijunzh**
Fix #12926. 

The current docs for creating a large `.npy` file using the memmap module is unclear. It will be more helpful if one can click a link to see the usage of `lib.format.open_memmap` function. I was not able to find a page or section on the online docs for `lib.format.open_memmap`. So, I added a link to the GitHub file for now.

- Add a link to `lib.format.open_memmap` function on GitHub
- Create a page or section under `lib.format` to detail the usage of `open_memmap` function.

### CODE REVIEW

**Comment by mhvk** [comment]
@lijunzh - I think the idea is good but the implementation not so much: this makes the text hard to parse for someone who just looks at the docstring interactively (e.g., in ipython, does `np.memmap?`).

But your PR brings up what I think is a bug: the `See also` section is supposed to automatically generate links in the documentation. As a random example of how it is supposed to look, see, e.g., [`np.sort`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.sort.html). I am not 100% sure, but think the problem arises because `open_memmap` itself does not have a separate documentation entry.

I think the problem is that there is no documentation generated for `numpy/lib/format.py` - this may be related to how the summary is generated in `doc/sources/reference/routines.io.rst`. To do it, one probably would want to define an `__all__` and make use the standard template.

p.s. Possibly more logical, though, might be to move `open_memmap` to `lib/npyio.py`...

**Comment by lijunzh** [comment]
@mhvk I totally agree with you that the implementation is, by no means, acceptable. I put it here as an initiative for a problem confused me and possibly some others. I am not yet familiar with the Numpy docs/repo, so I would love to get some advice before I dump in and try to fix the real problem. 

Yes, the `np.sort` example you quoted is the exact output I am looking for. There is one page for [`lib.format`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.lib.format.html?highlight=lib%20format#module-numpy.lib.format); yet, there is no specific info about `open_memmap`. There is only one this sentence: "Allows memory-mapping of the data. See open_memmep" (with the wrong spelling for `open_memmap`). I will look into the doc-generation process more and see if I can come up with a better solution. I think a page/section for `open_memmap` is necessary.

I like the idea of moving `open_memmap` to `lib.npyio`. However, I think we will want to keep the old API accessible in `lib.format` so that tools like https://github.com/deep-compute/diskarray will continue to work, at least for a version or two.

Last, here are some of my thoughts on `.npy` format. Even though many ([here](https://stackoverflow.com/questions/9619199/best-way-to-preserve-numpy-arrays-on-disk) and [there](https://www.reddit.com/r/MachineLearning/comments/30ui94/storage_formats_for_large_datasets/) ) have recommended HDF5 over NPY for a large array, my personal experience after working with both formats for more than three years suggest that NPY is still a better solution for Python. There are some excellent tools, like [hickle](https://github.com/telegraphic/hickle), and HDF5 is language independent. However, HDF5 has some [legitimate disadvantages](https://cyrille.rossant.net/moving-away-hdf5/). Since data scientists are, by large, Python programmers now, I see NPY as an outstanding solution for data scientists. This situation make it important for Numpy, as a fundamental computing tool in Python, to support a user-friendly interface for creating large NPY file on disk and populate it in real time. Here is a typical use case I see (maybe not possible yet):

```Python
data = some_iterator_to_spit_out_samples
with numpy_function_for_creating_npy_on_disk(path, **kwargs) as f:
    for i, sample in enumerate(data):
        f[i] = sample
        some_code_to_flush_array_to_disk
```

I will be happy to help improve the current documetation on `open_memmap` as well as implmenting possible new tools to make it user-friendly. Suggestions are welcome.

**Comment by mhvk** [comment]
@lijunzh - thanks for the detailed comments! I agree that the npy format has its place, and that it is very handy to work with it using memmap. 

I think for starters we should just try to get the documentation to include `open_memmap` - we can decide on whether or not to move it later (another possibility might be to support it as a class method on `memmap`). It would be great if you can help. I'm always somewhat confused by how sphinx works - a lot is done automatically - but I'd maybe start by giving `format.py` an `__all__ = ['open_memmap']` and then in `routines.io.rst` removing the `autosummary/minimal_module.rst` template line.

**Comment by lijunzh** [comment]
@mhvk Thanks for the positive feedback. I will start to get the docs straightened, and we can start from there later. I had some experience with sphinx before with [my little package](https://www.yews.info/docs). I will try to figure it out myself before I ask any silly questions.

**Comment by seberg** [comment]
Maybe we should first also mention `np.load` which has a memmap_mode kwarg, and already works fine in most cases when loading?

**Comment by lijunzh** [comment]
@seberg Agreed. We should reference the `memmap_mode` in `np.load` too. @mhvk and I were mostly talking about creating a `.npy` file using `memmap`; however, we should consider loading as well for the completeness.

**Comment by mhvk** [comment]
@seberg - I actually didn't know about that, so definitely a good idea to add a link to it in `np.memmap`,

File: numpy/core/memmap.py [hunk 1]
```diff
@@ -104,7 +104,7 @@ class memmap(ndarray):
 
     See also
     --------
-    lib.format.open_memmap : Create or load a memory-mapped ``.npy`` file.
+    `lib.format.open_memmap <https://github.com/numpy/numpy/blob/v1.16.1/numpy/lib/format.py>`_ : Create or load a memory-mapped ``.npy`` file.
```
Comment (charris) [inline numpy/core/memmap.py]: This link isn't the best, there should be a better way to do this and also keep the line at reasonable length. This is also not the only reference that needs fixing @mattip Suggestions? `format.py` seems to lack an `__all__`.

Comment (mattip) [reply numpy/core/memmap.py]: The link should be created automatically once the we link in the documentation in `lib.format`. Adding a link to github is wrong.

Comment (lijunzh) [reply numpy/core/memmap.py]: @charris and @mattip Thanks for the review. I will try to fix it today.

**Comment by mattip** [comment]
> in routines.io.rst removing the autosummary/minimal_module.rst template line.

This part of the review is still not done

**Comment by rossbar** [comment]
Since this is still open, perhaps doc improvements related to #5976 could be considered as well

**Comment by rossbar** [comment]
@lijunzh gentle ping - do you plan to continue working on this? I have a suggestion that I think would fix all the sphinx linking that I could push up here - just wanted to check with you first. If you don't have the time/bandwidth now no worries!

**Comment by rossbar** [comment]
I am going to close this as `lib.format.open_memmap` was added to the documentation in #16664. #16681 has been opened to track similar considerations for the rest of the functions in `lib.format`.
